File: scripts/verify_coordinates.py
#!/usr/bin/env python3

import json
import os
import sys

# 台北市和新北市的合理座標範圍
VALID_BOUNDS = {
    # 擴大範圍以涵蓋雙北地區
    'min_lat': 24.90,  # 最南（新店、中和）
    'max_lat': 25.30,  # 最北（淡水、北投）
    'min_lng': 121.35, # 最西（林口）
    'max_lng': 121.70  # 最東（南港、汐止）
}

def validate_coordinates(geojson_path):
    print(f"\n🔍 驗證檔案: {geojson_path}")
    
    try:
        with open(geojson_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        features = data.get('features', [])
        valid_count = 0
        invalid_count = 0
        out_of_bounds_points = []
        
        for index, feature in enumerate(features):
            coords = feature['geometry']['coordinates']
            lng = coords[0]
            lat = coords[1]
            name = feature['properties']['name']
            feature_type = feature['properties']['type']
            city = feature['properties'].get('city', '台北市')
            
            # 檢查座標是否在有效範圍內
            if (lat >= VALID_BOUNDS['min_lat'] and lat <= VALID_BOUNDS['max_lat'] and
                lng >= VALID_BOUNDS['min_lng'] and lng <= VALID_BOUNDS['max_lng']):
                valid_count += 1
                print(f"✅ [{index + 1}] {name} ({feature_type}) - {city}")
                print(f"   座標: [{lng}, {lat}]")
            else:
                invalid_count += 1
                issues = []
                
                if lat < VALID_BOUNDS['min_lat']:
                    issues.append('緯度過低')
                if lat > VALID_BOUNDS['max_lat']:
                    issues.append('緯度過高')
                if lng < VALID_BOUNDS['min_lng']:
                    issues.append('經度過低')
                if lng > VALID_BOUNDS['max_lng']:
                    issues.append('經度過高')
                
                out_of_bounds_points.append({
                    'index': index + 1,
                    'name': name,
                    'type': feature_type,
                    'city': city,
                    'coordinates': [lng, lat],
                    'issues': issues
                })
                
                print(f"❌ [{index + 1}] {name} ({feature_type}) - {city}")
                print(f"   座標: [{lng}, {lat}] - {', '.join(issues)}")
        
        print(f"\n📊 驗證結果:")
        print(f"   總點數: {len(features)}")
        print(f"   有效座標: {valid_count}")
        print(f"   無效座標: {invalid_count}")
        
        if out_of_bounds_points:
            print(f"\n⚠️  需要修正的座標:")
            for point in out_of_bounds_points:
                print(f"   • {point['name']} - {', '.join(point['issues'])}")
                
                # 提供建議座標
                suggested_lat = point['coordinates'][1]
                suggested_lng = point['coordinates'][0]
                
                if suggested_lat < VALID_BOUNDS['min_lat']:
                    suggested_lat = VALID_BOUNDS['min_lat'] + 0.01
                if suggested_lat > VALID_BOUNDS['max_lat']:
                    suggested_lat = VALID_BOUNDS['max_lat'] - 0.01
                if suggested_lng < VALID_BOUNDS['min_lng']:
                    suggested_lng = VALID_BOUNDS['min_lng'] + 0.01
                if suggested_lng > VALID_BOUNDS['max_lng']:
                    suggested_lng = VALID_BOUNDS['max_lng'] - 0.01
                
                print(f"     建議座標: [{suggested_lng}, {suggested_lat}]")
        
        return {
            'total': len(features),
            'valid': valid_count,
            'invalid': invalid_count,
            'out_of_bounds': out_of_bounds_points
        }
        
    except Exception as error:
        print(f"❌ 讀取檔案錯誤: {error}")
        return None

def validate_map_bounds():
    print('🗺️  台北城市儀表板 - 座標驗證工具')
    print('=======================================')
    print(f"有效座標範圍:")
    print(f"  緯度: {VALID_BOUNDS['min_lat']} ~ {VALID_BOUNDS['max_lat']}")
    print(f"  經度: {VALID_BOUNDS['min_lng']} ~ {VALID_BOUNDS['max_lng']}")
    
    files = [
        'Taipei-City-Dashboard-FE/public/mapData/market_events_taipei.geojson',
        'Taipei-City-Dashboard-FE/public/mapData/market_events_metrotaipei.geojson'
    ]
    
    total_stats = {
        'total': 0,
        'valid': 0,
        'invalid': 0,
        'files': 0
    }
    
    for file in files:
        if os.path.exists(file):
            result = validate_coordinates(file)
            if result:
                total_stats['total'] += result['total']
                total_stats['valid'] += result['valid']
                total_stats['invalid'] += result['invalid']
                total_stats['files'] += 1
        else:
            print(f"⚠️  檔案不存在: {file}")
    
    print(f"\n🎯 總體統計:")
    print(f"   檢查檔案: {total_stats['files']}")
    print(f"   總點數: {total_stats['total']}")
    print(f"   有效座標: {total_stats['valid']}")
    print(f"   無效座標: {total_stats['invalid']}")
    
    if total_stats['total'] > 0:
        success_rate = (total_stats['valid'] / total_stats['total']) * 100
        print(f"   成功率: {success_rate:.1f}%")
    
    if total_stats['invalid'] == 0:
        print(f"\n🎉 所有座標都在有效範圍內！地圖應該能正確顯示所有點位。")
    else:
        print(f"\n🔧 建議修正 {total_stats['invalid']} 個無效座標以確保完整顯示。")

if __name__ == "__main__":
    validate_map_bounds() 

File: cleanup-backup-20250601_075011/scripts/verify-coordinates.js
#!/usr/bin/env node

// 座標驗證腳本 - 檢查市集活動座標是否在合理範圍內
const fs = require('fs');
const path = require('path');

// 台北市和新北市的合理座標範圍
const VALID_BOUNDS = {
  // 擴大範圍以涵蓋雙北地區
  minLat: 24.90,  // 最南（新店、中和）
  maxLat: 25.30,  // 最北（淡水、北投）
  minLng: 121.35, // 最西（林口）
  maxLng: 121.70  // 最東（南港、汐止）
};

function validateCoordinates(geojsonPath) {
  console.log(`\n🔍 驗證檔案: ${geojsonPath}`);
  
  try {
    const data = JSON.parse(fs.readFileSync(geojsonPath, 'utf8'));
    const features = data.features || [];
    
    let validCount = 0;
    let invalidCount = 0;
    let outOfBoundsPoints = [];
    
    features.forEach((feature, index) => {
      const coords = feature.geometry.coordinates;
      const lng = coords[0];
      const lat = coords[1];
      const name = feature.properties.name;
      const type = feature.properties.type;
      const city = feature.properties.city || '台北市';
      
      // 檢查座標是否在有效範圍內
      if (lat >= VALID_BOUNDS.minLat && lat <= VALID_BOUNDS.maxLat &&
          lng >= VALID_BOUNDS.minLng && lng <= VALID_BOUNDS.maxLng) {
        validCount++;
        console.log(`✅ [${index + 1}] ${name} (${type}) - ${city}`);
        console.log(`   座標: [${lng}, ${lat}]`);
      } else {
        invalidCount++;
        outOfBoundsPoints.push({
          index: index + 1,
          name,
          type,
          city,
          coordinates: [lng, lat],
          issues: []
        });
        
        if (lat < VALID_BOUNDS.minLat) outOfBoundsPoints[outOfBoundsPoints.length - 1].issues.push('緯度過低');
        if (lat > VALID_BOUNDS.maxLat) outOfBoundsPoints[outOfBoundsPoints.length - 1].issues.push('緯度過高');
        if (lng < VALID_BOUNDS.minLng) outOfBoundsPoints[outOfBoundsPoints.length - 1].issues.push('經度過低');
        if (lng > VALID_BOUNDS.maxLng) outOfBoundsPoints[outOfBoundsPoints.length - 1].issues.push('經度過高');
        
        console.log(`❌ [${index + 1}] ${name} (${type}) - ${city}`);
        console.log(`   座標: [${lng}, ${lat}] - ${outOfBoundsPoints[outOfBoundsPoints.length - 1].issues.join(', ')}`);
      }
    });
    
    console.log(`\n📊 驗證結果:`);
    console.log(`   總點數: ${features.length}`);
    console.log(`   有效座標: ${validCount}`);
    console.log(`   無效座標: ${invalidCount}`);
    
    if (outOfBoundsPoints.length > 0) {
      console.log(`\n⚠️  需要修正的座標:`);
      outOfBoundsPoints.forEach(point => {
        console.log(`   • ${point.name} - ${point.issues.join(', ')}`);
        
        // 提供建議座標
        let suggestedLat = point.coordinates[1];
        let suggestedLng = point.coordinates[0];
        
        if (suggestedLat < VALID_BOUNDS.minLat) suggestedLat = VALID_BOUNDS.minLat + 0.01;
        if (suggestedLat > VALID_BOUNDS.maxLat) suggestedLat = VALID_BOUNDS.maxLat - 0.01;
        if (suggestedLng < VALID_BOUNDS.minLng) suggestedLng = VALID_BOUNDS.minLng + 0.01;
        if (suggestedLng > VALID_BOUNDS.maxLng) suggestedLng = VALID_BOUNDS.maxLng - 0.01;
        
        console.log(`     建議座標: [${suggestedLng}, ${suggestedLat}]`);
      });
    }
    
    return {
      total: features.length,
      valid: validCount,
      invalid: invalidCount,
      outOfBounds: outOfBoundsPoints
    };
    
  } catch (error) {
    console.error(`❌ 讀取檔案錯誤: ${error.message}`);
    return null;
  }
}

function validateMapBounds() {
  console.log('🗺️  台北城市儀表板 - 座標驗證工具');
  console.log('=======================================');
  console.log(`有效座標範圍:`);
  console.log(`  緯度: ${VALID_BOUNDS.minLat} ~ ${VALID_BOUNDS.maxLat}`);
  console.log(`  經度: ${VALID_BOUNDS.minLng} ~ ${VALID_BOUNDS.maxLng}`);
  
  const files = [
    'Taipei-City-Dashboard-FE/public/mapData/market_events_taipei.geojson',
    'Taipei-City-Dashboard-FE/public/mapData/market_events_metrotaipei.geojson'
  ];
  
  let totalStats = {
    total: 0,
    valid: 0,
    invalid: 0,
    files: 0
  };
  
  files.forEach(file => {
    if (fs.existsSync(file)) {
      const result = validateCoordinates(file);
      if (result) {
        totalStats.total += result.total;
        totalStats.valid += result.valid;
        totalStats.invalid += result.invalid;
        totalStats.files++;
      }
    } else {
      console.log(`⚠️  檔案不存在: ${file}`);
    }
  });
  
  console.log(`\n🎯 總體統計:`);
  console.log(`   檢查檔案: ${totalStats.files}`);
  console.log(`   總點數: ${totalStats.total}`);
  console.log(`   有效座標: ${totalStats.valid}`);
  console.log(`   無效座標: ${totalStats.invalid}`);
  console.log(`   成功率: ${((totalStats.valid / totalStats.total) * 100).toFixed(1)}%`);
  
  if (totalStats.invalid === 0) {
    console.log(`\n🎉 所有座標都在有效範圍內！地圖應該能正確顯示所有點位。`);
  } else {
    console.log(`\n🔧 建議修正 ${totalStats.invalid} 個無效座標以確保完整顯示。`);
  }
}

// 執行驗證
validateMapBounds(); 